core/lib/components/balena/sdk.js

```diff
@@ -418,4 +447,99 @@ module.exports = class BalenaSDK {
+	 * @param application the balena application name
+	 * @param directory the directory name containing the docker-compose for the application and the source files
+	*/
+	async pushReleaseToApp(application, directory){
+		await exec(
+			`balena push ${application} --source ${directory}`
+		  );
+		//check new commit of app
+		let commit = await this.balena.models.application.get(
+			application
+		)
+		.get("commit");
+
+		return commit
+	}
+	/** Waits until given servuces are all running on a device, on a given commit
+	 * @param uuid the uuid of the device
+	 * @param services an array of the service names
+	 * @param commit the release commit hash that services should be on
+	*/
+	async waitUntilServicesRunning(uuid, services, commit){
+		await utils.waitUntil(async () => {
+			let deviceServices = await this.balena.models.device.getWithServiceDetails(
+				uuid
+			  );
+			let running = false
+			running = services.every((service) => {
+				return (deviceServices.current_services[service][0].status === "Running") && (deviceServices.current_services[service][0].commit === commit)
+			})
+			return running;
+		}, false)
+	}
+
+	/** Executes the command in the targetted container of a device
+	 * @param command the command to be run
+	 * @param containerName the name of the service/container to run the command in
+	 * @param uuid the uuid of the device
+	*/
+	async executeCommandInContainer(command, containerName, uuid){
+		// get the container ID of container through balena engine
+		const containerId = await this.executeCommandInHostOS(
+			`balena ps --format "{{.Names}}" | grep ${containerName}`,
+			uuid
+		);
+
+		const stdout = await this.executeCommandInHostOS(
+			`balena exec ${containerId} ${command}`,
+			uuid
+		);
+
+		return stdout
+	}
+	/** Checks if device logs contain a string
+	 * @param uuid the uuid of the device
+	 * @param contains string to look for in the logs
+	 * @param start (optional) start the search from this log
+	 * @param start (optional) end the search at this log
+	*/
+	async checkLogsContain(uuid, contains, _start=null, _end=null){
+		let logs = await this.balena.logs.history(uuid)
+          .map((log) => {
+            return log.message;
+          });
+
+		let startIndex = (_start != null)? logs.indexOf(_start) : 0 
+		let endIndex = (_end != null)? logs.indexOf(_end) : (logs.length)
+		
+        let slicedLogs = logs.slice(startIndex, endIndex);
+
+        let pass = false;
+        slicedLogs.forEach((element) => {
+          if (element.includes(contains)) {
+            pass = true;
+          }
+        });
+
+		return pass
+	}
+
+	/** Returns the supervisor version on the device
+	 * @param uuid the uuid of the device
+	*/
+	async getSupervisorVersion(uuid){
+		let supervisor = await this.executeCommandInHostOS(
+		  `balena exec resin_supervisor cat package.json | grep version`,
+		  uuid
+		);
+		// we get something like - `"version": "12.3.5"`
```

Comment:
?

Reply:
I was getting the supervisor version by checking inside the supervisor container - I'm guessing there's a better way? xD

Reply:
Using the supervisor API requires the supervisor API key. Using the sdk to get device state only gets the version of the supervisor version as stored in the cloud at that moment - looking inside the actual supervisor container to get the version seemed like the most honest way to check its version, especially after it had just been purged and re downloaded

Reply:
Well, I mean why the comment is there. Do you mean to keep it there?

Reply:
I didn't check the approach in my first review. This seems fine. But, you can use the supervisor API as we just concluded from the device tree tests that calling the SV API from the host OS doesn't need the SV API key. Now that was in unmanaged balenaOS in local mode. Do check what happens when you are in managed balenaOS

Reply:
* This seems fine until they change resin-supervisor to balena_supervisor.